launcher/launcher.go

```diff
@@ -36,3 +63,805 @@ type NodeInfo struct {
+	}
+
+	instanceInfo.StoragePath, instanceInfo.StatePath, err = launcher.storageStateProvider.Setup(stateStrageParams)
+	if err != nil {
+		log.Errorf("Can't setup storage and state for instance: %v", err)
+
+		_ = launcher.uidPool.RemoveID(uid)
+
+		return instanceInfo, aoserrors.Errorf("Can't setup storage and state for instance: %v", err)
+	}
+
+	return instanceInfo, nil
+}
+
+func (launcher *Launcher) getNodesByStaticResources(
+	serviceInfo imagemanager.ServiceInfo, instanceInfo cloudprotocol.InstanceInfo,
+) (nodes []*nodeStatus, errStatus cloudprotocol.InstanceStatus) {
+	nodes = launcher.getNodeByRunner(serviceInfo.Config.Runner)
+	if len(nodes) == 0 {
+		log.Errorf("No appropriate node with runner: %s", serviceInfo.Config.Runner)
+
+		return nodes, createInstanceStatusFromInfo(instanceInfo, 0, serviceInfo.AosVersion,
+			cloudprotocol.InstanceStateFailed, fmt.Sprintf(
+				"No appropriate node with runner: %s", serviceInfo.Config.Runner))
+	}
+
+	nodes = launcher.getNodesByResources(nodes, serviceInfo.Config.Resources)
+	if len(nodes) == 0 {
+		log.Errorf("No appropriate node with resources: %v", serviceInfo.Config.Resources)
+
+		return nodes, createInstanceStatusFromInfo(instanceInfo, 0, serviceInfo.AosVersion,
+			cloudprotocol.InstanceStateFailed, fmt.Sprintf(
+				"No appropriate node with resourceser: %v", serviceInfo.Config.Resources))
+	}
+
+	nodes = launcher.getNodesByLabels(nodes, instanceInfo.Labels)
+	if len(nodes) == 0 {
+		log.Errorf("No appropriate node with labels %v", instanceInfo.Labels)
+
+		return nodes, createInstanceStatusFromInfo(instanceInfo, 0, serviceInfo.AosVersion,
+			cloudprotocol.InstanceStateFailed, fmt.Sprintf(
+				"No appropriate node with labels: %v", instanceInfo.Labels))
+	}
+
+	return nodes, errStatus
+}
+
+func (launcher *Launcher) getNodesByDevices(
+	nodes []*nodeStatus, desiredDevices []aostypes.ServiceDevice,
+) (newNodes []*nodeStatus) {
+	if len(desiredDevices) == 0 {
+		return slices.Clone(nodes)
+	}
+
+	for _, node := range nodes {
+		if len(node.availableDevices) == 0 {
+			continue
+		}
+
+		nodeAdded := true
+
+		for _, desiredDevice := range desiredDevices {
+			if !launcher.preallocateDevicesForNode(node, desiredDevice.Name) {
+				nodeAdded = false
+				break
+			}
+		}
+
+		if nodeAdded {
+			newNodes = append(newNodes, node)
+		}
+
+		for i := range node.availableDevices {
+			if nodeAdded {
+				node.availableDevices[i].allocatedCount = node.availableDevices[i].preallocateCount
+			} else {
+				node.availableDevices[i].preallocateCount = node.availableDevices[i].allocatedCount
+			}
+		}
+	}
+
+	return newNodes
+}
+
+func (launcher *Launcher) preallocateDevicesForNode(node *nodeStatus, deviseName string) bool {
+	for i, nodeDevice := range node.availableDevices {
+		if nodeDevice.name != deviseName {
+			continue
+		}
+
+		if nodeDevice.sharedCount == 0 {
+			return true
+		}
+
+		if nodeDevice.preallocateCount < nodeDevice.sharedCount {
```

Comment:
it's correct that we compare  sharedCount value but never decrement this value ?

Reply:
the shared count is stick configuration from unitConfig. In it initialised on sturtup